Replace virtualenv with Python's built-in venv

frewsxcv opened the pull request:
Since we require Python 3.5, we can just utilize Python's built-in `venv` tool for handling virtual environments.

frewsxcv committed:
Replace virtualenv with Python's built-in venv.

sagudev commented:
I wonder how this is done in mozilla now. Maybe they also use venv. Also it might be good to do mach sync before this.

delan commented on python/mach_bootstrap.py:
```diff
@@ -82,15 +82,12 @@
     for name in names:
+        print(name)
         path = find_executable(name)
+        print(path)
```
do we still need these prints?

delan commented on python/mach_bootstrap.py:
```diff
@@ -132,32 +129,18 @@ def _process_exec(args):
+        )
+        if not os.path.exists(virtualenv_path):
+            _process_exec([python, "-m", "venv", "--system-site-packages", virtualenv_path])
+        os.execv(venv_python_path, [venv_python_path] + sys.argv)
```
os.execv is known to misbehave on windows (python/cpython#101191, python/cpython#63323). can we use the same approach as activate_this.py, modifying PATH + VIRTUAL_ENV + sys.path + sys.real_prefix + sys.prefix? (see [these](https://stackoverflow.com/a/33637378) [answers](https://stackoverflow.com/a/33637414) for more details)

delan commented on python/mach_bootstrap.py:
```diff
@@ -191,6 +169,10 @@ def _activate_virtualenv(topdir):
 
 
+def in_venv():
+    return sys.prefix != sys.base_prefix
```
this check always returns true on NixOS, even when i’m outside a venv, which seems to be due to the way nixpkgs [installs pip and virtualenv by default](https://github.com/NixOS/nixpkgs/blob/92f31b5b7c33632c91f0b79204c8632388c47ece/pkgs/development/interpreters/python/wrapper.nix):

* sys.prefix = /nix/store/9p4gi4425qbrgn10qf2v21vyjr7q8l2m-python3-3.10.12-env
* sys.base_prefix = /nix/store/bc45k1n0pkrdkr3xa6w84w1xhkl1kkyp-python3-3.10.12

could we instead check for the presence of VIRTUAL_ENV in the environment?

delan reviewed (changes requested):
This looks good at a high level, but there are a couple of platform-specific problems.

sagudev commented:
Do note that mozjs's SM build sys might still require virtualenv.

mrobinson commented:
> Do note that mozjs's SM build sys might still require virtualenv.

It looks like Gecko switched from virtualenv to venv here: https://bugzilla.mozilla.org/show_bug.cgi?id=1766497

sagudev commented:
Then this should be done via mach sync. The first step would be to move mach to third_party and then do the sync.

jdm commented:
I think we may have forked their mach? I'll see if I can find those old discussions.

mrobinson committed:
Apply Delan's suggestions and make a couple small fixes

- Fix a tidy warning about directories that don't exist
- Use shutil instead of the redundant get_exec_path
- Miscellaneous cleanups

mrobinson commented:
I'd really like to land this and clear it out of the PR queue so I've gone ahead and updated the change. @delan, would you mind taking another look?

delan commented on python/mach_bootstrap.py:
```diff
@@ -120,15 +133,22 @@ def _activate_virtualenv(topdir):
-            _get_virtualenv_script_dir(),
-            "python"
-        )
+    if os.environ.get("VIRUTAL_ENV") != virtualenv_path:
```
typo: VIRUTAL_ENV → VIRTUAL_ENV

mrobinson replied:
Whoops. This is fixed in the latest branch.

delan reviewed (commented):
This fails for me with:

```
$ ./mach build -d
NOTE: Entering nix-shell etc/shell.nix
"/nix/store/x3m1f8qhs74lqz9mbfcc6v1wd5v060ql-python3-3.10.13-env/bin/python3.10" "-m" "pip" "install" "-I" "-r" "/cuffs/code/servo/python/requirements.txt" failed with error code 1:
Output:
Ignoring setuptools: markers 'python_version < "3.8"' don't match your environment
Ignoring dataclasses: markers 'python_version < "3.7"' don't match your environment
Collecting blessings==1.7
  Using cached blessings-1.7-py3-none-any.whl (18 kB)
Collecting distro==1.4
  Using cached distro-1.4.0-py2.py3-none-any.whl (17 kB)
Collecting mozinfo==1.2.1
  Using cached mozinfo-1.2.1-py2.py3-none-any.whl (7.1 kB)
Collecting mozlog==7.1.0
  Using cached mozlog-7.1.0-py2.py3-none-any.whl (69 kB)
Collecting setuptools==68.2.2
  Using cached setuptools-68.2.2-py3-none-any.whl (807 kB)
Collecting toml==0.9.2
  Using cached toml-0.9.2-py3-none-any.whl
Collecting flake8==3.8.3
  Using cached flake8-3.8.3-py2.py3-none-any.whl (72 kB)
Collecting pep8==1.5.7
  Using cached pep8-1.5.7-py2.py3-none-any.whl (37 kB)
Collecting pyflakes==2.2.0
  Using cached pyflakes-2.2.0-py2.py3-none-any.whl (66 kB)
Collecting ply==3.8
  Using cached ply-3.8-py2.py3-none-any.whl
Collecting colorama==0.3.7
  Using cached colorama-0.3.7-py2.py3-none-any.whl (19 kB)
Collecting boto3==1.28.84
  Using cached boto3-1.28.84-py3-none-any.whl (135 kB)
Collecting pyOpenSSL==23.0.0
  Using cached pyOpenSSL-23.0.0-py3-none-any.whl (57 kB)
Collecting PyGithub==1.58.1
  Using cached PyGithub-1.58.1-py3-none-any.whl (312 kB)
Collecting six==1.15
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting notify-py==0.3.42
  Using cached notify_py-0.3.42-py3-none-any.whl (649 kB)
Collecting flask
  Using cached flask-3.0.0-py3-none-any.whl (99 kB)
Collecting requests
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
Collecting types-requests
  Using cached types_requests-2.31.0.10-py3-none-any.whl (14 kB)
Collecting mozfile>=0.12
  Using cached mozfile-3.0.0-py2.py3-none-any.whl (8.2 kB)
Collecting mozterm
  Using cached mozterm-1.0.0-py2.py3-none-any.whl (3.4 kB)
Collecting pycodestyle<2.7.0,>=2.6.0a1
  Using cached pycodestyle-2.6.0-py2.py3-none-any.whl (41 kB)
Collecting mccabe<0.7.0,>=0.6.0
  Using cached mccabe-0.6.1-py2.py3-none-any.whl (8.6 kB)
Collecting s3transfer<0.8.0,>=0.7.0
  Using cached s3transfer-0.7.0-py3-none-any.whl (79 kB)
Collecting jmespath<2.0.0,>=0.7.1
  Using cached jmespath-1.0.1-py3-none-any.whl (20 kB)
Collecting botocore<1.32.0,>=1.31.84
  Using cached botocore-1.31.85-py3-none-any.whl (11.3 MB)
Collecting cryptography<40,>=38.0.0
  Using cached cryptography-39.0.2-cp36-abi3-manylinux_2_28_x86_64.whl (4.2 MB)
Collecting pynacl>=1.4.0
  Using cached PyNaCl-1.5.0-cp36-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.manylinux_2_24_x86_64.whl (856 kB)
Collecting deprecated
  Using cached Deprecated-1.2.14-py2.py3-none-any.whl (9.6 kB)
Collecting pyjwt[crypto]>=2.4.0
  Using cached PyJWT-2.8.0-py3-none-any.whl (22 kB)
Collecting loguru<=0.6.0,>=0.5.3
  Using cached loguru-0.6.0-py3-none-any.whl (58 kB)
Collecting jeepney
  Using cached jeepney-0.8.0-py3-none-any.whl (48 kB)
Collecting Werkzeug>=3.0.0
  Using cached werkzeug-3.0.1-py3-none-any.whl (226 kB)
Collecting itsdangerous>=2.1.2
  Using cached itsdangerous-2.1.2-py3-none-any.whl (15 kB)
Collecting blinker>=1.6.2
  Using cached blinker-1.7.0-py3-none-any.whl (13 kB)
Collecting Jinja2>=3.1.2
  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)
Collecting click>=8.1.3
  Using cached click-8.1.7-py3-none-any.whl (97 kB)
Collecting certifi>=2017.4.17
  Using cached certifi-2023.11.17-py3-none-any.whl (162 kB)
Collecting charset-normalizer<4,>=2
  Using cached charset_normalizer-3.3.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (142 kB)
Collecting urllib3<3,>=1.21.1
  Using cached urllib3-2.1.0-py3-none-any.whl (104 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.6-py3-none-any.whl (61 kB)
Collecting python-dateutil<3.0.0,>=2.1
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting urllib3<3,>=1.21.1
  Using cached urllib3-2.0.7-py3-none-any.whl (124 kB)
Collecting cffi>=1.12
  Using cached cffi-1.16.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (443 kB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.1.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)
Collecting wrapt<2,>=1.10
  Using cached wrapt-1.16.0-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (80 kB)
Collecting pycparser
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: toml, ply, pep8, mccabe, distro, colorama, wrapt, urllib3, six, setuptools, pyjwt, pyflakes, pycparser, pycodestyle, MarkupSafe, loguru, jmespath, jeepney, itsdangerous, idna, click, charset-normalizer, certifi, blinker, Werkzeug, types-requests, requests, python-dateutil, notify-py, mozterm, mozfile, Jinja2, flake8, deprecated, cffi, blessings, pynacl, mozlog, mozinfo, flask, cryptography, botocore, s3transfer, pyOpenSSL, PyGithub, boto3
Error:
ERROR: Could not install packages due to an OSError: [Errno 30] Read-only file system: '/nix/store/x3m1f8qhs74lqz9mbfcc6v1wd5v060ql-python3-3.10.13-env/lib/python3.10/site-packages/toml.py'
```

Looking into it.

mrobinson committed:
Fix typo in environment variable

delan committed:
fix bug where pip still tries to the wrong site-packages

delan reviewed (approved):
71872cb01b76b14d6f05b973f30558bd4952cf3b fixes the failure on NixOS by running pip with the venv’s python, rather than our original sys.executable.

mrobinson commented:
Thanks for the review and the fix.
